source/distributed/grid_tools.cc

```diff
@@ -0,0 +1,166 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+
+#include <deal.II/base/bounding_box.h>
+#include <deal.II/base/point.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/distributed/grid_tools.h>
+
+#ifdef DEAL_II_WITH_P4EST
+
+#include <vector>
+#include <cmath>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace GridTools
+  {
+    template < int dim, int spacedim>
+    std::vector<BoundingBox<spacedim> >
+    compute_locally_owned_bounding_box
+    (const parallel::distributed::Triangulation< dim, spacedim > &distributed_tria)
+    {
+      // In order to return the boundary points of the locally_owned cells correctly
+      // we need to handle the possibility of a second connected component.
+      //
+      // Assuming the triangulation is distributed following the morton-z curve order
+      // we know that, if there are two connected components, then the first and the
+      // last locally owned cell belong to different components.
+      //
+      // Thus we begin by looking for those cells which we call firstc and lastc
+
+      typename parallel::distributed::Triangulation< dim, spacedim >::active_cell_iterator
+      firstc = distributed_tria.begin_active();
+      typename parallel::distributed::Triangulation< dim, spacedim >::active_cell_iterator
+      lastc = distributed_tria.last_active();
+
+      while ( ! firstc->is_locally_owned() )
+        ++firstc;
+
+
+      while ( ! lastc->is_locally_owned() )
+        --lastc;
+
+      //Degenerate cases: only one owned cell or we're at level 0
+      if (lastc==firstc || firstc->level()*lastc->level()==0)
```

Comment:
I don't understand the || and the second condition. Doesn't that mean that it is enough that one of the cells is at level 0? Why do you need to consider only the cell firstc if the first expression is false?